File: backend/src/models/oracle.py
"""Pydantic models for Oracle feature."""

from typing import List, Optional, Literal, Any, Dict
from datetime import datetime
from pydantic import BaseModel, Field


class SourceReference(BaseModel):
    """Source citation from oracle retrieval."""
    path: str = Field(..., description="File path, note path, or thread ID")
    source_type: Literal["code", "vault", "thread", "repomap"] = Field(
        ..., description="Type of knowledge source"
    )
    line: Optional[int] = Field(None, description="Line number for code sources")
    snippet: Optional[str] = Field(None, description="Relevant excerpt (max 500 chars)")
    score: Optional[float] = Field(None, description="Relevance score (0.0-1.0)")
    metadata: Optional[Dict[str, Any]] = Field(None, description="Additional metadata")


class OracleRequest(BaseModel):
    """Request payload for oracle queries."""
    question: str = Field(..., min_length=1, max_length=2000, description="Natural language question")
    sources: Optional[List[Literal["vault", "code", "threads"]]] = Field(
        None, description="Knowledge sources to query (None = all)"
    )
    explain: bool = Field(False, description="Include retrieval traces for debugging")
    model: Optional[str] = Field(None, description="Override LLM model (e.g., 'anthropic/claude-3.5-sonnet')")
    thinking: bool = Field(False, description="Enable thinking mode (append :thinking suffix)")
    max_tokens: int = Field(16000, ge=1000, le=100000, description="Maximum tokens for context assembly")


class OracleStreamChunk(BaseModel):
    """Server-sent event chunk for streaming responses."""
    type: Literal["thinking", "content", "source", "done", "error"] = Field(
        ..., description="Chunk type"
    )
    content: Optional[str] = Field(None, description="Text content for thinking/content chunks")
    source: Optional[SourceReference] = Field(None, description="Source citation for source chunks")
    tokens_used: Optional[int] = Field(None, description="Total tokens used (done chunk only)")
    model_used: Optional[str] = Field(None, description="Model used (done chunk only)")
    error: Optional[str] = Field(None, description="Error message (error chunk only)")
    metadata: Optional[Dict[str, Any]] = Field(None, description="Additional metadata")


class OracleResponse(BaseModel):
    """Response payload for non-streaming oracle queries."""
    answer: str = Field(..., description="Synthesized answer")
    sources: List[SourceReference] = Field(default_factory=list, description="Source citations")
    tokens_used: Optional[int] = Field(None, description="Total tokens used")
    model_used: Optional[str] = Field(None, description="Model that generated the response")
    retrieval_traces: Optional[Dict[str, Any]] = Field(None, description="Retrieval debug info (if explain=True)")


class ConversationMessage(BaseModel):
    """A single message in conversation history."""
    role: Literal["user", "assistant", "system"] = Field(..., description="Message role")
    content: str = Field(..., description="Message content")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Message timestamp")
    sources: Optional[List[SourceReference]] = Field(None, description="Sources for assistant messages")


class ConversationHistoryResponse(BaseModel):
    """Response payload for conversation history."""
    messages: List[ConversationMessage] = Field(default_factory=list, description="Conversation history")
    session_id: Optional[str] = Field(None, description="Session identifier")
    compressed: bool = Field(False, description="Whether history has been compressed")
    token_count: Optional[int] = Field(None, description="Approximate token count")

File: frontend/src/types/oracle.ts
export type SourceType = 'vault' | 'code' | 'threads';
export type StreamEventType = 'status' | 'thinking' | 'content' | 'source' | 'tool_call' | 'tool_result' | 'done' | 'error';

/**
 * Oracle API request payload
 */
export interface OracleRequest {
  question: string;
  sources?: SourceType[];
  explain?: boolean;
  max_results?: number;
  model?: string;            // Override LLM model (e.g., 'deepseek/deepseek-chat')
  thinking?: boolean;        // Enable thinking mode
  max_tokens?: number;       // Maximum tokens for context assembly
  context_id?: string;       // Context tree node ID for conversation continuity
}

/**
 * Retrieval result from any knowledge source
 */
export interface RetrievalResult {
  content: string;
  source_type: SourceType;
  source_path: string;
  score: number;
  metadata?: Record<string, unknown>;
}

/**
 * Stream event chunks from Oracle SSE endpoint
 */
export interface OracleStreamChunk {
  type: StreamEventType;
  message?: string;          // for status events
  content?: string;          // for content/thinking tokens
  source?: RetrievalResult;  // for source events
  tokens_used?: number;      // for done event
  duration_ms?: number;      // for done event
  model_used?: string;       // for done event (matches backend)
  context_id?: string;       // for done event - context node ID for next request
  error?: string;            // for error events
  tool_call?: {              // for tool_call events
    id: string;
    name: string;
    arguments: string;
    status?: string;
  };
  tool_call_id?: string;     // for tool_result events - associates result with tool call
  tool_result?: string;      // for tool_result events
}

/**
 * Slash command definition
 */
export interface SlashCommand {
  name: string;
  description: string;
  shortcut?: string;
  handler: () => void | Promise<void>;
}

/**
 * Tool call with execution result
 */
export interface ToolCallInfo {
  id: string;
  name: string;
  arguments: string;
  status?: 'pending' | 'running' | 'completed' | 'error';
  result?: string;
}

/**
 * Oracle conversation message (extends ChatMessage)
 */
export interface OracleMessage {
  role: 'user' | 'assistant';
  content: string;
  timestamp: string;
  thinking?: string;         // Optional thinking/reasoning content
  sources?: RetrievalResult[];
  tool_calls?: ToolCallInfo[]; // Tool calls made during response
  model?: string;
  is_error?: boolean;
}

/**
 * Oracle session settings
 */
export interface OracleSettings {
  model?: string;
  sources?: SourceType[];
  show_thinking?: boolean;
  show_sources?: boolean;
  max_results?: number;
}
